### PR TITLE
evalengine: Add support for enum and set

### PR DESCRIPTION
**Pull request by dbussink**
The evalengine currently doesn't handle enum and set types properly. The comparison function always returns 0 at the moment and we don't consider ordering of elements etc.

Here we add two new native types to the evalengine for set and enum and instantiate those appropriately. We also ensure we can compare them correctly.

In case we don't have the schema information with the values, we do a best effort case of depending on the string representation. This is not correct always of course, but at least makes equality comparison work for those cases and only ordering is off in that scenario.

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/15782

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by vmg — COMMENTED** [review]
Very nice addition! I have one big piece of feedback, which is about the `Values []string` type that has been wired up throughout the codebase. It's heavy! It's super heavy with 24-bytes + backing storage + all the string backing storage.

Clearly, neither `ENUM` or `SET` is a commonly used feature (as evidenced by the fact that we've gotten away without implementing support for it so far). Hence, I think this warrants spending some time optimizing the storage of these values so they don't weight down the rest of the evalengine whenever they're not used (which will be 99% of the time). This is particularly important for the evalengine's internal and external `Type`, which are currently pretty well packed (yey!) and they would almost triple in size after adding the `[]string` values.

I'm not 100% sure of what's the ideal way to optimize this. This is how I would try though: start with a `type EnumSetValues string` and store `values *EnumSetValues` everywhere where they must be kept. The full set of values is serialized into a single immutable string, and the `EnumSetValues` would have helpers to iterate each value individually / get their index/etc without having to understand how they're serialized.

This (or something close to this) feels like the cheapest way to pass along the set of values without requiring the 24 bytes of the slice. Give this a go, or otherwise hit me up and we can pair on it. :wink:

**Review by vmg — APPROVED** [review]
After discussing in Slack, the sweet middle ground is probably `*[]string` to hide the slice behind a pointer without having to bother with a packed serialization format. LGTM!

**Review by mattlord — APPROVED** [review]
Thanks! I'll merge this into https://github.com/vitessio/vitess/pull/15723 tomorrow.

One tangentially related thing that I also realized in working on that is that for ENUMs we can get the length of the column when building vevents in unit tests this way:
```
int64(len(col.Type.EnumValues) + 1)
```

There is no equivalent for Set in the parser.

File: go/vt/vttablet/tabletmanager/vreplication/replicator_plan.go [hunk 1]
```diff
@@ -303,7 +303,7 @@ func (tp *TablePlan) isOutsidePKRange(bindvars map[string]*querypb.BindVariable,
 
 		rowVal, _ := sqltypes.BindVariableToValue(bindvar)
 		// TODO(king-11) make collation aware
```
[the remark this answers is not in the document]
Comment (dbussink) [reply go/vt/vttablet/tabletmanager/vreplication/replicator_plan.go:305]: This one still passes in `Unknown` for the collation, so I don't think we can remove it (yet)?

**Comment by dbussink** [comment]
> One tangentially related thing that I also realized in working on that is that for ENUMs we can get the length of the column when building vevents in unit tests this way:
> 
> ```
> int64(len(col.Type.EnumValues) + 1)
> ```
> 
> There is no equivalent for Set in the parser.

@mattlord A `set` should always be backed under the hood storage wise by a single `uint64`. A set is then stored as a bitmap. So you can always make that assumption then.

**Pull request merged**